pkg/loop/internal/chain_reader.go

```diff
@@ -0,0 +1,148 @@
+		}
+	case JSONEncodingVersion2:
+		bytes, err = jsonv2.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
+		}
+	case CBOREncodingVersion:
+		enco := cbor.CoreDetEncOptions()
+		enco.Time = cbor.TimeRFC3339Nano
+		var enc cbor.EncMode
+		enc, err = enco.EncMode()
+		if err != nil {
+			return nil, err
+		}
+		bytes, err = enc.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported encoding version %d for data %v", version, data)
+	}
+
+	return &pb.VersionedBytes{Version: uint32(version), Data: bytes}, nil
+}
+
+func decodeVersionedBytes(res any, vData *pb.VersionedBytes) error {
+	var err error
+	switch vData.Version {
+	case JSONEncodingVersion1:
+		err = jsonv1.Unmarshal(vData.Data, res)
+	case JSONEncodingVersion2:
+		err = jsonv2.Unmarshal(vData.Data, res)
+	case CBOREncodingVersion:
+		err = cbor.Unmarshal(vData.Data, res)
+	default:
+		return fmt.Errorf("unsupported encoding version %d for versionedData %v", vData.Version, vData.Data)
+	}
+
+	if err != nil {
+		return fmt.Errorf("%w: %w", types.InvalidTypeError{}, err)
+	}
+	return nil
+}
+
+func (c *chainReaderClient) GetLatestValue(ctx context.Context, bc types.BoundContract, method string, params, retVal any) error {
+	versionedParams, err := encodeVersionedBytes(params, ParamsCurrentEncodingVersion)
+	if err != nil {
+		return err
+	}
+
+	boundContract := pb.BoundContract{Name: bc.Name, Address: bc.Address, Pending: bc.Pending}
+
+	reply, err := c.grpc.GetLatestValue(ctx, &pb.GetLatestValueRequest{Bc: &boundContract, Method: method, Params: versionedParams})
+	if err != nil {
+		return unwrapClientError(err)
+	}
+
+	return decodeVersionedBytes(retVal, reply.RetVal)
+}
+
+var _ pb.ChainReaderServer = (*chainReaderServer)(nil)
+
+type chainReaderServer struct {
+	pb.UnimplementedChainReaderServer
+	impl types.ChainReader
+}
+
+func (c *chainReaderServer) GetLatestValue(ctx context.Context, request *pb.GetLatestValueRequest) (*pb.GetLatestValueReply, error) {
+	var bc types.BoundContract
+	bc.Name = request.Bc.Name[:]
+	bc.Address = request.Bc.Address[:]
+	bc.Pending = request.Bc.Pending
+
+	params := &map[string]any{}
+
+	if err := decodeVersionedBytes(params, request.Params); err != nil {
+		return nil, err
+	}
+
+	retVal := &map[string]any{}
+	err := c.impl.GetLatestValue(ctx, bc, request.Method, params, retVal)
+	if err != nil {
+		return nil, err
+	}
+
+	encodedRetVal, err := encodeVersionedBytes(retVal, RetvalCurrentEncodingVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.GetLatestValueReply{RetVal: encodedRetVal}, nil
+}
+
+func unwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		switch s.Message() {
+		case types.InvalidTypeError{}.Error():
+			return types.InvalidTypeError{}
+		case types.FieldNotFoundError{}.Error():
+			return types.FieldNotFoundError{}
```

Comment:
I think we can simplify this code (and the other uses) if we define these a different way:
```suggestion
		case types.InvalidTypeError:
			return types.InvalidTypeError
		case types.FieldNotFoundError:
			return types.FieldNotFoundError
```
```go
type errChainReader string

func (e errChainReader) Error() string { return e}

const (
	ErrInvalidType = errChainReader("asdf")
	ErrFieldNotFound = errChainReader("foo-bar")
)
```

Reply:
Not sure how this can simplify the unwrap. Maybe you meant this?:

`func unwrapClientError(err error) error {
	if s, ok := status.FromError(err); ok {
		return ErrChainReader(s)
	}
	return err
}`

ErrChainReader must be exported then though

Reply:
nice